Add County Divisions and Parishes Sub-Modules

AWearring opened the pull request:
## What does this change?

Resolves:  https://github.com/localgovdrupal/localgov_elections/issues/148 & https://github.com/localgovdrupal/localgov_elections/issues/149

1. Sub-module created to add a new Boundary Source Provider for County Divisions based upon the `LocalGov Elections Reporting - ONS Wards 2023` sub-module using [Ward to LAD to County to County Electoral Division (May 2024) Lookup for EN](https://open-geography-portalx-ons.hub.arcgis.com/datasets/ons::ward-to-lad-to-county-to-county-electoral-division-may-2024-lookup-for-en/explore) and [County Electoral Division (May 2023) Boundaries EN BFE](https://geoportal.statistics.gov.uk/datasets/ons::county-electoral-division-may-2023-boundaries-en-bfe-2/explore?showTable=true)
2. Sub-module created to add a new Boundary Source Provider for Parishes based upon the `LocalGov Elections Reporting - ONS Wards 2023` sub-module using [Parishes (December 2024) Names and Codes in EW](https://geoportal.statistics.gov.uk/datasets/c6d944797d1f4de19a6f1dedcedce1f5_0/explore?showTable=true) and [Parishes and Non Civil Parished Areas (December 2024) Boundaries EW BFC](https://geoportal.statistics.gov.uk/datasets/4148460530ec427693ddc31094b9164c_0/explore?showTable=true)

## How to test

### County Divisions

1. Enable sub-module
3. Navigate to Boundary Sources `/admin/structure/boundary-source`
4. Add a provider - select "ONS Divisions 2024"
5. Enter valid County Code (and optionally a valid District Code) and submit
6. Create a new election
7. Add areas - select newly created Boundary Source
8. Select some or all of the County Divisions
9. Check new areas have been created with defined boundaries/maps

###  Parishes

1. Enable sub-module
3. Navigate to Boundary Sources `/admin/structure/boundary-source`
4. Add a provider - select "ONS Parishes 2024"
5. Enter valid District (LAD) Code and submit
6. Create a new election
7. Add areas - select newly created Boundary Source
8. Select some or all of the Parishes
9. Check new areas have been created with defined boundaries/maps

## How can we measure success?

1. Divisions can be created for County Council Elections
2. Parishes can be created for Parish Council Elections

AWearring committed:
Merge branch '1.x' of https://github.com/localgovdrupal/localgov_elections into 1.x

willguv commented:
This is amazing, thanks @AWearring. Do we need to add something to the readme?

https://github.com/localgovdrupal/localgov_elections/blob/1.x/docs/index.md

AWearring committed:
Translatable strings must not begin or end with white spaces

AWearring commented:
@willguv An update to https://github.com/localgovdrupal/localgov_elections/blob/1.x/README.md would be more appropriate

AWearring committed:
Update ReadMe

AWearring committed:
Fix Translateable strings

AWearring committed:
Fix coding standards

dedavidson commented:
@AWearring I will try and review today but more likely Monday.

AWearring commented:
@dedavidson Thanks - I have been given some feedback/comments from the Big Blue Door team to improve the code, so there will be some updates later today. In light of this I'll convert it to a draft until I've made the changes.

AWearring committed:
Updates following Big Blue Door Dev comments

AWearring committed:
Add Parishes Sub-module

AWearring committed:
Fixed: Global constants should not be used, move it to a class or interface

AWearring committed:
More coding standards

AWearring committed:
Update README.md

Ad "LocalGov Elections Reporting ONS Parishes 2024 - Boundary source provider for Office of National Statistics 2024 Parishes" to list of sub-modules

AWearring commented:
@tran-bigbluedoor Helga reviewed and approved these as a custom module for `west-lindsey-bbd` - could you review it as a sub-module in `localgov_elections`?

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Form/OnsTwentyFourDivisionsDownloadForm.php:
```diff
@@ -0,0 +1,148 @@
+  public function __construct(Client $http_client, RequestStack $request) {
+    $this->httpClient = $http_client;
+    $this->request = $request;
+  }
```
Can we use constructor promotion here? See: https://stitcher.io/blog/constructor-promotion-in-php-8

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Form/OnsTwentyFourDivisionsDownloadForm.php:
```diff
@@ -0,0 +1,148 @@
+   *
+   * @var \Drupal\localgov_elections\BoundaryProviderInterface
+   */
+  protected $plugin;
```
Can we move this class variable to the top, before the `create` function?

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Form/OnsTwentyFourDivisionsDownloadForm.php:
```diff
@@ -0,0 +1,148 @@
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
```
Can we put all the code to generate the opts to another func.
```suggestion
  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
      $opts = $this->getAreasToDownload();
      $form = [ ...]
      return $form;
  }
```

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Form/OnsTwentyFourDivisionsDownloadForm.php:
```diff
@@ -0,0 +1,148 @@
+    $lad = $this->plugin->getConfiguration()['lad'];
+    $cty = $this->plugin->getConfiguration()['cty'];
+    $url = self::URL_SERVICES_LU;
+    if ($lad && $cty) {
+      $url = $url . 'where=' . self::URL_WHERE_CTY . $cty . '%27%20AND%20' . self::URL_WHERE_LAD . $lad . '%27&' . self::URL_FIELDS_CED;
+    }
+    elseif (!$lad && $cty) {
+      $url = $url . 'where=' . self::URL_WHERE_CTY . $cty . '%27&' . self::URL_FIELDS_CED;
+    }
+    else {
+      $url = $url . 'where=' . self::URL_WHERE_LAD . $lad . '%27&' . self::URL_FIELDS_CED;
+    }
```
If both `$lad` and `$city` are empty?

ekes replied:
Also here and elsewhere these are getting passed to Guzzle. It has params options. So no need to generate (encoded) strings, you can build an array (which for the logic might be cleaner) and pass them as an argument to Guzzle so does the work https://docs.guzzlephp.org/en/stable/request-options.html#query

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Form/OnsTwentyFourDivisionsDownloadForm.php:
```diff
@@ -0,0 +1,148 @@
+    else {
+      $url = $url . 'where=' . self::URL_WHERE_LAD . $lad . '%27&' . self::URL_FIELDS_CED;
+    }
+    $response = $this->httpClient->get($url);
```
Can we wrap all httpClient request in a `try ... catch`

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Plugin/BoundaryProvider/OnsTwentyFourDivisions.php:
```diff
@@ -0,0 +1,302 @@
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   Messenger service.
+   */
+  public function __construct(
+    array $configuration,
+    $plugin_id,
+    $plugin_definition,
+    Client $http_client,
+    entityTypeManagerInterface $entity_type_manager,
+    MessengerInterface $messenger,
+  ) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->httpClient = $http_client;
+    $this->nodeStorage = $entity_type_manager->getStorage('node');
+    $this->paragraphStorage = $entity_type_manager->getStorage('paragraph');
+    $this->messenger = $messenger;
+  }
```
Same here, can we use cosntructor promotion?

AWearring replied:
Hi @tran-bigbluedoor
I'm struggling with this - the following throws an error
```
public function __construct(
    public array $configuration,
    public string $plugin_id,
    public string $plugin_definition,
    public Client $http_client,
    public entityTypeManagerInterface $entity_type_manager,
    public MessengerInterface $messenger,
  ) {}
```

AWearring replied:
@tran-bigbluedoor  How about this?
```public function __construct(
    $configuration,
    $plugin_id,
    $plugin_definition,
    public Client $http_client,
    public entityTypeManagerInterface $entity_type_manager,
    public MessengerInterface $messenger,
  ) {
    parent::__construct($configuration, $plugin_id, $plugin_definition);
    $this->nodeStorage = $entity_type_manager->getStorage('node');
    $this->paragraphStorage = $entity_type_manager->getStorage('paragraph');
  }
```

ekes replied:
I'm with Matt Glaman on the use of the constructor to get the storage from the DI https://mglaman.dev/blog/dependency-injection-anti-patterns-drupal not that I'd die on that hill.

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Plugin/BoundaryProvider/OnsTwentyFourDivisions.php:
```diff
@@ -0,0 +1,302 @@
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+
+    $lad_url = self::URL_LAD;
+    $cty_url = self::URL_CTY;
```
```suggestion
```

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Plugin/BoundaryProvider/OnsTwentyFourDivisions.php:
```diff
@@ -0,0 +1,302 @@
+      '#title' => $this->t('Local Authority County Code (CTY24CD)'),
+      '#maxlength' => 1000,
+      '#default_value' => $this->configuration['cty'] ?? "",
+      '#description' => $this->t('County code. You can find this <a href="@url">here</a>. Use the value from the CTY24CD column.', ['@url' => $cty_url]),
```
```suggestion
      '#description' => $this->t('County code. You can find this <a href="@url">here</a>. Use the value from the CTY24CD column.', ['@url' => self::URL_CTY]),
```

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Plugin/BoundaryProvider/OnsTwentyFourDivisions.php:
```diff
@@ -0,0 +1,302 @@
+      '#title' => $this->t('Local Authority District Code (LAD23CD)'),
+      '#maxlength' => 1000,
+      '#default_value' => $this->configuration['lad'] ?? "",
+      '#description' => $this->t('Local Authority District code. You can find this <a href="@url">here</a>. Use the value from the LAD23CD column.', ['@url' => $lad_url]),
```
```suggestion
      '#description' => $this->t('Local Authority District code. You can find this <a href="@url">here</a>. Use the value from the LAD23CD column.', ['@url' => self::URL_LAD]),
```

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Plugin/BoundaryProvider/OnsTwentyFourDivisions.php:
```diff
@@ -0,0 +1,302 @@
+    $gis_url = self::URL_SERVICES_CED . 'where=' . self::URL_WHERE_CED . $list . '&' . self::URL_FIELDS_GEOJSON;
+    $matched_features = [];
+    $response = $this->httpClient->get($gis_url);
+    if ($response->getStatusCode() == 200) {
+      $body = $response->getBody()->getContents();
+
+      $json_decoded = json_decode($body, TRUE);
+      $num_to_match = count($ids);
+      $num_matched = 0;
+      if ($response->getStatusCode() == 200) {
```
Duplicate  `($response->getStatusCode() == 200`

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Plugin/BoundaryProvider/OnsTwentyFourDivisions.php:
```diff
@@ -0,0 +1,302 @@
+          'title' => $election_node->getTitle() . ' - ' . $name,
+        ]
+      );
+      $area->save();
```
Should we clear all the `localgov_area_vote` node for that election, before creating a new set?

AWearring replied:
I'll add this as an extra issue as I think it is wiser than this sub-module: https://github.com/localgovdrupal/localgov_elections/issues/152

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Plugin/BoundaryProvider/OnsTwentyFourDivisions.php:
```diff
@@ -0,0 +1,302 @@
+
+    $boundaries = $this->fetchBoundaryInformation($vals);
+    $election = $form_values['localgov_election'];
+    $election_node = $this->nodeStorage->load($election);
```
Can we check if the `$election_node` is an instanceof NodeInterface first

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Plugin/BoundaryProvider/OnsTwentyFourDivisions.php:
```diff
@@ -0,0 +1,302 @@
+    elseif (!$lad && $cty) {
+      $url = $url . 'where=' . self::URL_WHERE_CTY . $cty . '%27&' . self::URL_FIELDS_JSON;
+    }
+    else {
+      $url = $url . 'where=' . self::URL_WHERE_LAD . $lad . '%27&' . self::URL_FIELDS_JSON;
+    }
```
Same here, `$lad` can be empty here

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Plugin/BoundaryProvider/OnsTwentyFourDivisions.php:
```diff
@@ -0,0 +1,302 @@
+    if ($response->getStatusCode() == 200) {
+      $body = $response->getBody()->getContents();
+      $json_decoded = json_decode($body, TRUE);
+      $features = $json_decoded['features'];
```
```suggestion
      $features = $json_decoded['features'] ?? [];
```

AWearring committed:
Updates to Divisions

AWearring committed:
Conding standards

AWearring committed:
Updates to Parishes

AWearring committed:
coding standards

AWearring committed:
more coding standards

AWearring commented:
Not sure what I need to do to resolve the [Static Analysis checks](https://github.com/localgovdrupal/localgov_elections/actions/runs/15026300798/job/42227911932?pr=151), most of them are nothing to do with this PR and the ones that do are common to the rest of the module.

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Form/OnsTwentyFourDivisionsDownloadForm.php:
```diff
@@ -0,0 +1,155 @@
+        '#title' => $this->t('Areas to download'),
+        '#type' => 'tableselect',
+        '#header' => ['area' => $this->t('Area')],
+        '#options' => &$opts,
```
```suggestion
        '#options' => $opts,
```

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Form/OnsTwentyFourDivisionsDownloadForm.php:
```diff
@@ -0,0 +1,155 @@
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getAreasToDownload() {
```
Please add the correct phpDoc for this function, also the return type.

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Plugin/BoundaryProvider/OnsTwentyFourDivisions.php:
```diff
@@ -0,0 +1,296 @@
+    $configuration,
+    $plugin_id,
+    $plugin_definition,
+    public Client $http_client,
+    public entityTypeManagerInterface $entity_type_manager,
```
```suggestion
    protected Client $httpClient,
    protected entityTypeManagerInterface $entityTypeManager,
```

AWearring replied:
Helga (BBD) requested that the case type should be consistent so snake or camel. 
Therefore keeping them in snake case so they are consistent with OnsTwentyFourDivisionsDownloadForm.php

ekes replied:
Drupal coding standards that LGD follows are: camel case for class properties so $entityTypeManager here. If it's a method variable then it's $entity_type_manager. So when you do a constructor old way you get `__construct($entity_type_manager) { $this->entityTypeManager = $entity_type_manager; }` but in this case it's going straight to a class property.

[edit: actually you might be allowed to use camel case for variables if you are consistent, but you can't use snake case for properties; generally practice is as written above]

AWearring replied:
OK will update to Camel case :-)

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Form/OnsTwentyFourDivisionsDownloadForm.php:
```diff
@@ -0,0 +1,155 @@
+   *   Messenger service.
+   */
+  public function __construct(
+    public Client $http_client,
+    public RequestStack $request,
+    public MessengerInterface $messenger,
```
```suggestion
    protected Client $http_client,
    protected RequestStack $request,
    protected MessengerInterface $messenger,
```

AWearring committed:
Update after comments

AWearring committed:
Coding Standards

AWearring committed:
More coding standards

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_divisions/src/Plugin/BoundaryProvider/OnsTwentyFourDivisions.php:
```diff
@@ -0,0 +1,296 @@
+    $plugin_id,
+    $plugin_definition,
+    protected Client $httpClient,
+    protected entityTypeManagerInterface $entity_type_manager,
```
```suggestion
    protected entityTypeManagerInterface $entityTypeManager,
```

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_parishes/src/Plugin/BoundaryProvider/OnsTwentyFourParishes.php:
```diff
@@ -0,0 +1,269 @@
+  ) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->nodeStorage = $entityTypeManager->getStorage('node');
+    $this->paragraphStorage = $entityTypeManager->getStorage('paragraph');
```
Where do you use this `paragraphStorage`?

AWearring replied:
Doesn't appear to be used at all, it is refernced in OnsTwentyThreeWards.php too and not used there either

tran-bigbluedoor replied:
Then please remove this variable @AWearring . I am wondering if we should remove `nodeStorage` as well. We can always load a node with  `$this-> entityTypeManager->getStorage('node')->load($id)`

AWearring committed:
Remove redundant variables

AWearring committed:
Removed unused use statement

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_parishes/src/Plugin/BoundaryProvider/OnsTwentyFourParishes.php:
```diff
@@ -0,0 +1,252 @@
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+
+    $lad_url = self::URL_LAD;
+
```
```suggestion
```

tran-bigbluedoor commented on modules/localgov_elections_ons_twenty_four_parishes/src/Plugin/BoundaryProvider/OnsTwentyFourParishes.php:
```diff
@@ -0,0 +1,252 @@
+      '#title' => $this->t('Local Authority District Code (LAD24CD)'),
+      '#maxlength' => 1000,
+      '#default_value' => $this->configuration['lad'] ?? "",
+      '#description' => $this->t('Local Authority District code. You can find this <a href="@url">here</a>. Use the value from the LAD24CD column.', ['@url' => $lad_url]),
```
```suggestion
      '#description' => $this->t('Local Authority District code. You can find this <a href="@url">here</a>. Use the value from the LAD24CD column.', ['@url' => self::URL_LAD]),
```

AWearring committed:
Removed unnecessary variable

finnlewis commented:
Thanks for this @AWearring  and @tran-bigbluedoor 

We'll merge this and add another issue to fix the phpstan failure to get the phpunit tests running again.

willguv commented:
@AWearring love this, can I write about it in the next product update newsletter please?

Were you able to update the elections readme to mention this, how to install and use?

AWearring commented:
@willguv As it isn't getting release until the phpstan failures have been addressed it should probably wait until then - when is the next newsletter due to go out?

They are listed as optional sub-modules in the Elections readme which explains how to enable them and both sub-modules have a readme to explain usage :)

willguv commented:
@AWearring we generally send it the first week of the month so have a bit of time left. Or could include in the next one. How fixable are the issues? We could point a bit of dev time that way

AWearring commented:
@willguv Looks like most of the fails are down to not having return types specified on functions - I'll look at adding those and see where it sits after they've been added
